src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -2153,14 +2445,50 @@ new_open_pthread(const char *pathname, int oflags, ...)
 	return rc;
 }
 
+/* Search a fd in fd hash table. Remove it in case it is found. Also free the fake fd.
+ * Return true if fd is found. Return false if fd is not found.
+ */
+static bool
+remove_fd_compatible(int real_fd)
+{
+	struct ht_fd *fd_ht_obj;
+	d_list_t     *rlink;
+
+	rlink = d_hash_rec_find(fd_hash, &real_fd, sizeof(int));
+	if (rlink == NULL)
+		return false;
+
+	fd_ht_obj = fd_obj(rlink);
+	/* close fake fd */
+	if (fd_ht_obj->fake_fd >= FD_DIR_BASE)
+		free_dirfd(fd_ht_obj->fake_fd - FD_DIR_BASE);
+	else
+		free_fd(fd_ht_obj->fake_fd - FD_FILE_BASE, false);
+	/* remove fd from hash table */
+	d_hash_rec_decref(fd_hash, rlink);
```

Comment:
Same comment as above, could `rec_delete` work here?

Also - and I've not looked into this bug could the free_dirfd/free_fd calls be made in the hash table `free` callback?  That would seem to make more sense logically but would probably impact other code using this hash table as well.

Reply:
d_hash_rec_decref() is replaced by d_hash_rec_delete_at(). Thank you!
I would prefer the current form. free_dirfd and free_fd are for fake fd. Here we are removing a hash table record of a real fd.

Reply:
fd_rec_free() is not invoked after d_hash_rec_decref() is replaced by d_hash_rec_delete_at().
With d_hash_rec_decref(), the reference count is decreased to 0 and fd_rec_free() is triggered.

Reply:
Ok, lets leave this as it is for now.  It may be the `D_HASH_FT_EPHEMERAL` flag.

Reply:
d_hash_rec_delete_at() is reverted back to d_hash_rec_decref().